receiver/prometheusreceiver/internal/transaction.go

```diff
@@ -71,14 +85,98 @@ type transaction struct {
 	bufBytes []byte
 }
 
-var emptyScopeID scopeID
+// ScopeIdentifier represents an identifier for a metric scope, which can include
+// just the basic scope information or also include scope attributes.
+type ScopeIdentifier interface {
```

Comment:
I would rather define a hash function on a metric scope if we can, and use maps of uint64.  String appending is very expensive--especially compared to hashing.  Maybe we could add it to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5404f7c2434387d8b68f3be6010a72fded822d85/pkg/pdatautil/hash.go?

Reply:
Map lookups with a uint64 is MUCH faster than with a data structure, although collisions are possible.  We could do the same for the ResourceMetrics lookup if we made a hash function for the resource.